### PR TITLE
DAOS-12255 control: Correct hugepage affinity with multi-engine-per-NUMA

### PR DESCRIPTION
**Pull request by tanabarr**
When multiple engines coexist on the same NUMA node (and there is an
equal number of engines per NUMA node), divide the total number of
hugepages proportionally across NUMA nodes when allocating.

Features: control

Required-githooks: true

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/server/server_utils.go [hunk 1]
```diff
@@ -208,20 +208,23 @@ func getFabricNetDevClass(cfg *config.Server, fis *hardware.FabricInterfaceSet)
...
-			nodes = []string{nn}
-			break
+		nodeMap[int(ec.Storage.NumaNodeIndex)] += 1
+	}
+
+	var lastCount int
+	nodes := make([]int, 0, len(engineCfgs))
+	for k, v := range nodeMap {
+		if lastCount != 0 && v != lastCount {
+			log.Errorf("different number of engines assigned to each NUMA node (%+v),"+
```
Comment (kjacque) [inline src/control/server/server_utils.go:221]: Seems like more of a Notice to me

File: src/control/server/server_utils_test.go [hunk 2]
```diff
@@ -555,6 +556,35 @@ func TestServer_prepBdevStorage(t *testing.T) {
 			expMemSize:      16384,
 			expHugePageSize: 2,
 		},
+		"4 engines; hugepage alloc across numa 0,1": {
```
Comment (kjacque) [inline src/control/server/server_utils_test.go:559]: Would be interesting to add a test for the case where there is an uneven distribution of engines.

File: src/control/server/server_utils.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/server_utils.go:221]: It is an error condition, having a different number of engines running on each NUMA will cause problems so I think it should be raised as an error in the log but for extraneous circumstances we want to allow the server to start in this case. As an aside, it was previously a Notice but wasn't being printed for some reason when the log mask was set to DEBUG.

File: src/control/server/server_utils_test.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/server_utils_test.go:559]: On inspection decided that it should be an error if engine configs are imbalanced across numa nodes (unless all engines are on a single node). Added unit test case for uneven distribution as requested.

File: src/control/server/server_utils.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/server_utils.go]: I changed this to return an error.

**Review by kjacque — DISMISSED** [review]

File: src/control/server/server_utils.go [hunk 3]
```diff
@@ -291,7 +298,9 @@ func prepBdevStorage(srv *server, iommuEnabled bool) error {
 		// allocation as some overhead may result in one or two being unavailable.
 		prepReq.HugePageCount = srv.cfg.NrHugepages / len(numaNodes)
 		prepReq.HugePageCount += common.ExtraHugePages
-		prepReq.HugeNodes = strings.Join(numaNodes, ",")
+		prepReq.HugeNodes = strings.Trim(
+			strings.Join(strings.Split(fmt.Sprint(numaNodes), " "), ","),
+			"[]")
```
Comment (mjmac) [inline src/control/server/server_utils.go:303]: 🤨 ... You're tying this functionality to the current implementation of how the language prints an integer slice. Is it likely to change? No. But I don't think the language compatibility guarantee extends to `fmt.Sprint()` output.

It looks like the only place you're using the `numaNodes` value returned by `getEngineNUMANodes()` is here, so why not just generate the string slice in that function while you're iterating over the nodes anyhow?

Comment (tanabarr) [reply src/control/server/server_utils.go:303]: that's what it's used to do and I will change back, I was going back on forth on this myself

Comment (tanabarr) [reply src/control/server/server_utils.go]: done

**Review by kjacque — APPROVED** [review]

**Comment by tanabarr** [comment]
single known CI failure:
- TelemetryIOLatency DAOS-12013

**Review by mjmac — APPROVED** [review]

**Pull request merged**